### PR TITLE
Introduce beartype & fix types

### LINKED ISSUE #202: feat: Use beartyping
Use beartyping to type check code.

### PR DESCRIPTION
**Pull request by st--**
Please check the type of change your PR introduces:

- [x] Bugfixes identified by added type checking
- [x] Some backwards-incompatible changes to fix behaviour
- [x] Code style update (formatting, renaming)

## What is the current behavior?

No runtime type checking.

Resolves #202 

## What is the new behavior?

Runtime type checking.

### CODE REVIEW

**Comment by st--** [comment]
Currently stuck on this issue: https://github.com/beartype/beartype/issues/226#issuecomment-1511019847

**Comment by st--** [comment]
I've resolved the issue https://github.com/beartype/beartype/issues/233 by removing the `from __future__ import annotations` and replacing forward-references with strings instead. Now back to fixing actual type mismatches...

File: tests/test_kernels/test_approximations.py [hunk 1]
```diff
@@ -130,12 +130,12 @@ def test_exactness(kernel):
 )
 def test_value_error(kernel):
     with pytest.raises(ValueError):
-        RFF(kernel(), num_basis_fns=10)
+        RFF(base_kernel=kernel(), num_basis_fns=10)
```
Comment (st--) [inline tests/test_kernels/test_approximations.py:133]: A better solution would be Python 3.10's KW_ONLY fields for the base classes

File: tests/test_kernels/test_non_euclidean.py [hunk 2]
```diff
@@ -25,6 +26,7 @@
 config.update("jax_enable_x64", True)
 
 
+@pytest.mark.skip("Graph kernel is broken")
```
Comment (st--) [inline tests/test_kernels/test_non_euclidean.py:29]: Couldn't figure out how to get the shapes/types to match up. Looks like a weird special-case that doesn't quite follow the general kernel API. Please check separately.

File: tests/test_kernels/test_nonstationary.py [hunk 3]
```diff
@@ -64,7 +64,7 @@ def test_initialization(self, fields: dict, dim: int) -> None:
         assert is_dataclass(self.kernel)
 
         # Input fields as JAX arrays
-        fields = {k: jnp.array([v]) for k, v in fields.items()}
+        fields = {k: jnp.array(v) for k, v in fields.items()}
```
Comment (st--) [inline tests/test_kernels/test_nonstationary.py:67]: backwards-incompatible change due to using proper scalar fields for e.g. variance

File: tests/test_kernels/test_stationary.py [hunk 4]
```diff
@@ -82,7 +82,7 @@ def test_initialization(self, fields: dict, dim: int) -> None:
         assert is_dataclass(self.kernel)
 
         # Input fields as JAX arrays
-        fields = {k: jnp.array([v]) for k, v in fields.items()}
+        fields = {k: jnp.array(v) for k, v in fields.items()}
```
Comment (st--) [inline tests/test_kernels/test_stationary.py:85]: backwards-incompatible change due to using proper scalar fields for e.g. variance

File: tests/test_mean_functions.py [hunk 5]
```diff
@@ -13,28 +14,28 @@ def test_abstract() -> None:
...
     assert isinstance(mf, AbstractMeanFunction)
-    assert (mf(jax.numpy.array([1.0])) == constant).all()
-    assert (mf(jax.numpy.array([2.0, 3.0])) == constant).all()
+    assert (mf(jnp.array([[1.0]])) == jnp.array([constant])).all()
+    assert (mf(jnp.array([[2.0, 3.0]])) == jnp.array([constant])).all()
     assert (
-        jax.vmap(mf)(jax.numpy.array([[1.0], [2.0]]))
-        == jax.numpy.array([constant, constant])
+        mf(jnp.array([[1.0], [2.0]]))
+        == jnp.array([constant, constant])
```
Comment (st--) [inline tests/test_mean_functions.py:36]: The `Constant` mean function doesn't quite work how you would expect from the base class (simply operate on a single data point, then we vmap over X); instead, it takes in the whole dataset. Should this become an issue to make consistent with kernel's `__call__` later?

File: tests/test_objectives.py [hunk 6]
```diff
@@ -23,15 +23,15 @@ def build_data(num_datapoints: int, num_dims: int, key, binary: bool):
             0.5
             * jnp.sign(
                 jnp.cos(
-                    3 * x[:, 1].reshape(-1, 1)
+                    3 * x[:, 0].reshape(-1, 1)
```
Comment (st--) [inline tests/test_objectives.py:26]: drive-by fix: the previous code shouldn't have worked for num_dims=1 ... :thinking:

File: tests/test_variational_families.py [hunk 7]
```diff
@@ -44,12 +45,16 @@ def test_abstract_variational_family():
         AbstractVariationalFamily()
 
     # Create a dummy variational family class with abstract methods implemented.
+    class DummyPosterior:
+        @property
+        def __class__(self) -> type: return AbstractPosterior
```
Comment (st--) [inline tests/test_variational_families.py:50]: the right way of mocking for beartype

File: gpjax/utils.py
Comment (st--) [inline gpjax/utils.py:1]: should we rename it gpjax/types.py ?

File: gpjax/utils.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+from beartype.typing import Union
+from jaxtyping import Bool, UInt32, Int, Float, Array
+from jax.random import KeyArray as JAXKeyArray
+
+OldKeyArray = UInt32[Array, "2"]
+KeyArray = Union[OldKeyArray, JAXKeyArray]  # for compatibility regardless of enable_custom_prng setting
```
Comment (st--) [inline gpjax/utils.py:21]: nb- not tested with the different enable_custom_prng settings

File: gpjax/kernels/computations/base.py [hunk 9]
```diff
@@ -15,15 +15,15 @@
...
+from beartype.typing import Any
 
 from jax import vmap
 from jaxtyping import Array, Float
 
 from gpjax.linops import (DenseLinearOperator, DiagonalLinearOperator,
                           LinearOperator)
 
-Kernel = Any
+Kernel = Any  # TODO: FIX
```
Comment (st--) [inline gpjax/kernels/computations/base.py:26]: this doesn't quite seem right. should we just use a string forward reference instead?

File: gpjax/kernels/__init__.py [hunk 10]
```diff
@@ -14,8 +14,9 @@
 # ==============================================================================
 
 """JaxKern."""
-from .approximations import RFF
 from .base import AbstractKernel, ProductKernel, SumKernel
+
+from .approximations import RFF
```
Comment (st--) [inline gpjax/kernels/__init__.py:19]: Has to come _after_ the `.base` imports so we don't get forward-reference issues!

File: gpjax/gaussian_distribution.py [hunk 11]
```diff
@@ -179,11 +181,11 @@ def _sample_n(self, key: KeyArray, n: int) -> Float[Array, "n N"]:
 
         return vmap(affine_transformation)(Z)
 
-    def sample(self,seed: KeyArray, sample_shape: Tuple[int, int]):  # pylint: disable=useless-super-delegation
-      """See `Distribution.sample`."""
-      return self._sample_n(seed, sample_shape[0])
+    def sample(self, seed: KeyArray, sample_shape: Tuple[int, ...]):  # pylint: disable=useless-super-delegation
+        """See `Distribution.sample`."""
+        return self._sample_n(seed, sample_shape[0])  # TODO this looks weird, why ignore the second entry?
```
Comment (st--) [inline gpjax/gaussian_distribution.py:186]: Is this a bug that should become its own issue?

File: gpjax/dataset.py [hunk 12]
```diff
@@ -84,7 +83,7 @@ def out_dim(self) -> int:
         return self.y.shape[1]
 
 
-def _check_shape(X: Float[Array, "N D"], y: Float[Array, "N Q"]) -> None:
+def _check_shape(X: Optional[Float[Array, "..."]], y: Optional[Float[Array, "..."]]) -> None:
```
Comment (st--) [inline gpjax/dataset.py:86]: Here I removed the beartype-shape checking. Could instead simply remove the `_check_shape` function...

File: gpjax/base/param.py [hunk 13]
```diff
@@ -13,12 +13,11 @@
 # limitations under the License.
 # ==============================================================================
 
-from __future__ import annotations
```
Comment (st--) [inline gpjax/base/param.py:16]: this caused beartype issues (https://github.com/beartype/beartype/issues/233)

File: gpjax/__init__.py [hunk 14]
```diff
@@ -13,19 +13,21 @@
...
-from .mean_functions import Constant, Zero
-from .objectives import (ELBO, CollapsedELBO, ConjugateMLL,
-                         LogPosteriorDensity, NonConjugateMLL)
-from .variational_families import (CollapsedVariationalGaussian,
-                                   ExpectationVariationalGaussian,
-                                   NaturalVariationalGaussian,
-                                   VariationalGaussian,
-                                   WhitenedVariationalGaussian)
+from jaxtyping import install_import_hook
+with install_import_hook("gpjax", "beartype.beartype"):
```
Comment (st--) [inline gpjax/__init__.py:17]: do we _want_ to push beartype onto end users?

File: gpjax/base/module.py [hunk 15]
```diff
@@ -31,7 +30,9 @@
                               PyTreeCheckpointer, PyTreeCheckpointHandler,
                               RestoreArgs, SaveArgs)
 from simple_pytree import Pytree, static_field
-from typing_extensions import Self
```
Comment (st--) [inline gpjax/base/module.py:34]: didn't seem to be compatible with beartype

File: gpjax/fit.py [hunk 16]
```diff
@@ -33,7 +33,7 @@
 def fit(
```
Comment (st--) [inline gpjax/fit.py:33]: NB: should `fit()` always jit the code itself ? should we make an issue for that?

File: gpjax/dataset.py (hunk 12 above)
Comment (thomaspinder) [reply gpjax/dataset.py:86]: Is there any reason not to remove `_check_shape` now we use beartype? If not, then let's do it.

File: gpjax/fit.py (hunk 16 above)
Comment (thomaspinder) [reply gpjax/fit.py:33]: I think `fit` should just accept a callable. Let's leave it as is for now and, once merged, we can open a PR to resolve #235

File: gpjax/gaussian_distribution.py (hunk 11 above)
Comment (thomaspinder) [reply gpjax/gaussian_distribution.py:186]: Yes. Done - #236

File: gpjax/kernels/base.py [hunk 17]
```diff
@@ -32,8 +37,8 @@
 class AbstractKernel(Module):
     """Base kernel class."""
 
-    compute_engine: AbstractKernelComputation = static_field(DenseKernelComputation)
-    active_dims: List[int] = static_field(None)
+    compute_engine: Type[AbstractKernelComputation] = static_field(DenseKernelComputation)
```
Comment (thomaspinder) [inline gpjax/kernels/base.py:40]: Why do we need `Type` here?

File: gpjax/kernels/base.py [hunk 18]
```diff
@@ -46,36 +51,36 @@ def cross_covariance(self, x: Float[Array, "N D"], y: Float[Array, "M D"]):
...
 
         Args:
-            x (Float[Array, "N D"]): The matrix or vector that is to be sliced.
+            x (Float[Array, "... D"]): The matrix or vector that is to be sliced.
         Returns:
-            Float[Array, "N S"]: A sliced form of the input matrix.
+            Float[Array, "... Q"]: A sliced form of the input matrix.
         """
-        return x[..., self.active_dims]
+        return x[..., self.active_dims] if self.active_dims is not None else x
```
Comment (thomaspinder) [inline gpjax/kernels/base.py:62]: Does this resolve #237?

File: gpjax/kernels/computations/base.py (hunk 9 above)
Comment (thomaspinder) [reply gpjax/kernels/computations/base.py:26]: Yes.

File: gpjax/likelihoods.py [hunk 19]
```diff
@@ -146,14 +143,14 @@ def predict(self, dist: tfd.Distribution) -> tfd.Distribution:
         return self.link_function(mean / jnp.sqrt(1.0 + variance))
 
 
-def inv_probit(x: Float[Array, "N 1"]) -> Float[Array, "N 1"]:
+def inv_probit(x: Float[Array, "*N"]) -> Float[Array, "*N"]:
```
Comment (thomaspinder) [inline gpjax/likelihoods.py:146]: What does the `*` in `*N` do?

File: gpjax/linops/dense_linear_operator.py [hunk 20]
```diff
@@ -107,7 +103,7 @@ def __mul__(self, other: float) -> LinearOperator:
 
         return DenseLinearOperator(matrix=self.matrix * other)
 
-    def _add_diagonal(self, other: DiagonalLinearOperator) -> LinearOperator:
+    def _add_diagonal(self, other: "gpjax.linops.diagonal_linear_operator.DiagonalLinearOperator") -> LinearOperator:
```
Comment (thomaspinder) [inline gpjax/linops/dense_linear_operator.py:106]: Why do we need `gpjax.linops.diagonal_linear_operator` here? Can we not just have `DiagonalLinearOperator`?

File: gpjax/linops/linear_operator.py [hunk 21]
```diff
@@ -35,8 +30,9 @@
 NestedT = Union[T, Iterable["NestedT"], Mapping[Any, "NestedT"]]
 
 # Nested types.
+DTypes = Union[Type[jnp.float32], Type[jnp.float64], Type[jnp.int32], Type[jnp.int64]]
 ShapeT = TypeVar("ShapeT", bound=NestedT[Tuple[int, ...]])
-DTypeT = TypeVar("DTypeT", bound=NestedT[jnp.dtype])
+DTypeT = TypeVar("DTypeT", bound=NestedT[DTypes])
```
Comment (thomaspinder) [inline gpjax/linops/linear_operator.py:35]: Could these be moved into `linops.utils` to be consistent with other types?

File: gpjax/linops/linear_operator.py [hunk 22]
```diff
@@ -60,58 +56,58 @@ def ndim(self) -> int:
...
         raise NotImplementedError
 
-    def __rmul__(self, other: float) -> LinearOperator:
+    def __rmul__(self, other: ScalarFloat) -> "LinearOperator":
         """Reimplimentation of multiply linear operator by scalar."""
         return self.__mul__(other)
 
     @abc.abstractmethod
-    def _add_diagonal(self, other: DiagonalLinearOperator) -> LinearOperator:
+    def _add_diagonal(self, other: "gpjax.linops.diagonal_linear_operator.DiagonalLinearOperator") -> "LinearOperator":
```
Comment (thomaspinder) [inline gpjax/linops/linear_operator.py:97]: Do we need `gpjax.linops.diagonal_linear_operator`?

File: gpjax/utils.py
Comment (thomaspinder) [reply gpjax/utils.py:1]: Yes. Or `typing`?

File: tests/test_gps.py [hunk 23]
```diff
@@ -229,13 +231,13 @@ def test_prior_sample_approx(num_datapoints, kernel, mean_function):
         p.sample_approx(-1, key)
     with pytest.raises(ValueError):
         p.sample_approx(0, key)
-    with pytest.raises(ValueError):
+    with pytest.raises((ValueError, beartype.roar.BeartypeCallHintParamViolation)):
```
[the remark this answers is not in the document]
Comment (thomaspinder) [reply tests/test_gps.py:234]: I'd prefer to remove `_check` functions? It feels cleaner. Is there a draw back to pushing beartype onto the user?

File: tests/test_kernels/test_approximations.py (hunk 1 above)
Comment (thomaspinder) [reply tests/test_kernels/test_approximations.py:133]: I would love to depend on 3.10. I worry that it's too restrictive though. Maybe this warrants more discussion.

File: tests/test_kernels/test_non_euclidean.py (hunk 2 above)
Comment (thomaspinder) [reply tests/test_kernels/test_non_euclidean.py:29]: Could you raise an issue for this? I'm not sure I follow entirely.

File: tests/test_mean_functions.py (hunk 5 above)
Comment (thomaspinder) [reply tests/test_mean_functions.py:36]: Yes, the mean functions in general could do with being updated to be consistent with the kernel object.

**Review by thomaspinder — CHANGES_REQUESTED** [review]
Nice work @st--. 

A few minor comments and points of clarification.

File: gpjax/__init__.py (hunk 14 above)
Comment (st--) [reply gpjax/__init__.py:17]: hm, mainly in some places we might have specified stricter types (array dtypes/shapes) than is strictly required by the code, so some code might have run fine if only beartype hadn't intervened. also, the explicit _check_shape error messages might be a bit more informative than the generic beartype ones. I don't know if there's also some more interaction with the jaxtyping `@jaxtyped` decorator.

File: gpjax/likelihoods.py (hunk 19 above)
Comment (st--) [reply gpjax/likelihoods.py:146]: like in `*args`, it gobbles up arbitrary dimensions. but beartype itself doesn't check the two are consistent

File: gpjax/linops/dense_linear_operator.py (hunk 20 above)
Comment (st--) [reply gpjax/linops/dense_linear_operator.py:106]: no, because it's in a different module, so you need to have the fully-qualified module name for beartype to be able to find the reference

File: gpjax/linops/linear_operator.py (hunk 21 above)
Comment (st--) [reply gpjax/linops/linear_operator.py:35]: it's not linops-specific so I would actually favour moving it into something like gpjax.types

File: gpjax/linops/linear_operator.py (hunk 22 above)
Comment (st--) [reply gpjax/linops/linear_operator.py:97]: yes, because it's a reference to a class in a different module

File: gpjax/kernels/base.py (hunk 17 above)
Comment (st--) [reply gpjax/kernels/base.py:40]: Because it's a class, not an instance.  The type of `DenseKernelComputation` is `Type[DenseKernelComputation]`. Compare to `DenseKernelComputation()`, whose type is `DenseKernelComputation`.

File: gpjax/dataset.py (hunk 12 above)
Comment (st--) [reply gpjax/dataset.py:86]: Actually, beartype only checks arguments individually, and does not check consistency of dimensions across multiple arguments/return values, so might be better to keep the `_check_shape` (so e.g. "N1 D" and "N2 D" shapes get flagged as not-matching).

File: gpjax/utils.py
Comment (st--) [reply gpjax/utils.py:1]: "types" feels more natural to me in principle, but it's stdlib `typing` and `jaxtyping`, so let's go with that.

File: tests/test_kernels/test_non_euclidean.py (hunk 2 above)
Comment (st--) [reply tests/test_kernels/test_non_euclidean.py:29]: Should've been more specific; I don't think the kernel itself is necessarily broken. Just the shape annotations really didn't match the actual use, and I couldn't (didn't want to) get my head around it and decided to just skip the test instead.

Maybe the thing to do instead, for now, is simply remove all the shape annotations on GraphKernel ?

File: tests/test_gps.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (st--) [reply tests/test_gps.py:234]: See https://github.com/JaxGaussianProcesses/GPJax/pull/230#discussion_r1176047677

File: tests/test_mean_functions.py (hunk 5 above)
Comment (st--) [reply tests/test_mean_functions.py:36]: opened #238

File: gpjax/utils.py
Comment (st--) [reply gpjax/utils.py:1]: done

File: gpjax/kernels/computations/base.py (hunk 9 above)
Comment (st--) [reply gpjax/kernels/computations/base.py]: done

File: gpjax/kernels/base.py (hunk 18 above)
Comment (st--) [reply gpjax/kernels/base.py:62]: Don't know, do we have a MRE ready?

File: gpjax/dataset.py (hunk 12 above)
Comment (st--) [reply gpjax/dataset.py:86]: Actually actually, jaxtyping itself checks those, I just had a bug in my toy example trying it out. But the error messages are still more verbose/less precise than the ones emitted by _check_shape :disappointed:

File: tests/test_kernels/test_non_euclidean.py (hunk 2 above)
Comment (thomaspinder) [reply tests/test_kernels/test_non_euclidean.py:29]: Understood. I'm fine with removing the shape annotations for now.

**Comment by st--** [comment]
Going through the notebooks, another issue: `Float[Array, "..."]` specifically matches JAX-arrays, but not numpy arrays. How do JAX-functions handle numpy-arrays? Do they implicitly convert them? Should we therefore relax type hints on all user-facing parts (not just Dataset but also e.g. predict()) to allow numpy arrays?

File: gpjax/dataset.py (hunk 12 above)
Comment (thomaspinder) [reply gpjax/dataset.py:86]: OK. I'm guessing there's no easy way for us to customise the error messages that Beartype throws?

Comment (thomaspinder) [reply gpjax/dataset.py:86]: If the answer to my above question is no, then let's keep `_check_shape`.

Comment (thomaspinder) [reply gpjax/dataset.py:86]: It looks like this is a *very* hot topic right now: https://github.com/google/jaxtyping/issues/6.

File: gpjax/__init__.py (hunk 14 above)
Comment (thomaspinder) [reply gpjax/__init__.py:17]: So if we didn't push beartype onto the end-user, then it would just be a testing utility for the package? I could easily imagine how any overly rigid beartype assertions could be annoying for an end user.

**Comment by thomaspinder** [comment]
> Going through the notebooks, another issue: `Float[Array, "..."]` specifically matches JAX-arrays, but not numpy arrays. How do JAX-functions handle numpy-arrays? Do they implicitly convert them? Should we therefore relax type hints on all user-facing parts (not just Dataset but also e.g. predict()) to allow numpy arrays?

Yes, JAX will implicitly convert them. So calling `posterior(np_array, D)` will produce the same result as `posterior(jnp.asarray(np_array), D)`. I think it'd place less friction on the user if we relax the type hint to accomodate Numpy arrays.

File: tests/test_kernels/test_non_euclidean.py (hunk 2 above)
Comment (st--) [reply tests/test_kernels/test_non_euclidean.py]: i've actually managed to fix the annotations to what's happening, but it still seems wrong

File: gpjax/__init__.py (hunk 14 above)
Comment (st--) [reply gpjax/__init__.py:17]: Yeah! Though on the other hand, it could help us discover more bugs not handled by the tests...

File: tests/test_kernels/test_non_euclidean.py (hunk 2 above)
Comment (st--) [reply tests/test_kernels/test_non_euclidean.py]: See #239

File: gpjax/dataset.py [hunk 24]
```diff
@@ -31,8 +30,8 @@ class Dataset(Pytree):
         y (Optional[Float[Array, "N Q"]]): Output data.
     """
 
-    X: Optional[Float[Array, "N D"]] = None
-    y: Optional[Float[Array, "N Q"]] = None
+    X: Optional[Union[Num[Array, "N D"], Num[Array, "..."]]] = None
+    y: Optional[Union[Float[Array, "N Q"], Float[Array, "..."]]] = None
```
Comment (st--) [inline gpjax/dataset.py:34]: should this also be
```suggestion
    y: Optional[Union[Num[Array, "N Q"], Num[Array, "..."]]] = None
```
for e.g. classification?

File: examples/graph_kernels.pct.py [hunk 25]
```diff
@@ -85,9 +85,9 @@
 
 true_kernel = gpx.GraphKernel(
     laplacian=L,
-    lengthscale=jnp.array([2.3]),
-    variance=jnp.array([3.2]),
-    smoothness=jnp.array([6.1]),
+    lengthscale=jnp.array(2.3),
+    variance=jnp.array(3.2),
+    smoothness=jnp.array(6.1),
```
Comment (st--) [inline examples/graph_kernels.pct.py:90]: NB- should it be possible as a user to simply pass numpy floats, e.g.
```suggestion
    lengthscale=2.3,
    variance=3.2,
    smoothness=6.1,
```
?

File: gpjax/__init__.py (hunk 14 above)
Comment (st--) [reply gpjax/__init__.py:17]: For example, a bunch of them were only uncovered by the notebooks. But I guess we could add beartype there as well, and then use that as a way to suggest to users that they _should_ use e.g. beartype (and report when one of our type hints is wrong)?

File: gpjax/kernels/nonstationary/linear.py [hunk 26]
```diff
@@ -17,7 +17,9 @@
 
 import jax.numpy as jnp
 import tensorflow_probability.substrates.jax.bijectors as tfb
-from jaxtyping import Array, Float
+from gpjax.typing import Array
+from jaxtyping import Float
+from gpjax.typing import ScalarFloat
```
Comment (thomaspinder) [inline gpjax/kernels/nonstationary/linear.py:22]: Can we merge lines 20 and 22 please? I see this occurs for most of the kernel modules.

File: gpjax/linops/linear_operator.py (hunk 21 above)
Comment (thomaspinder) [reply gpjax/linops/linear_operator.py:36]: Sure, let's do that.

File: gpjax/__init__.py (hunk 14 above)
Comment (thomaspinder) [reply gpjax/__init__.py:17]: That's a good way to position it. Let's push it onto users then. If it becomes an issue, we can always walk it back with little major disruption.

**Review by thomaspinder — APPROVED** [review]
All looks good to me. There's a small issue with imports, but I'm happy to merge as I've setup a pre-commit action on `docs_update` that will fix this. I will rebase that branch and correct.

If you're happy with the proposed approach of pushing beartype onto the user, then we can merge this.

File: gpjax/kernels/nonstationary/linear.py (hunk 26 above)
Comment (st--) [reply gpjax/kernels/nonstationary/linear.py:22]: Yeah, that's because I didn't want to touch two dozen files by hand and used `sed` for fixing the imports. I would prefer _not_ to merge those lines manually (a lot more effort because it's not always exactly the same so can't as easily write a `sed` line for it), instead, I would suggest tools like `isort` for consistent import order (similar to `black` for code formatting). But I would leave that for a separate PR that does nothing else other than "visual" changes (so e.g. one can ignore it in git blame).

Comment (thomaspinder) [reply gpjax/kernels/nonstationary/linear.py:22]: Yep, on `docs_update` that included in the pre-commit, so let's leave that job for then.

File: gpjax/__init__.py (hunk 14 above)
Comment (st--) [reply gpjax/__init__.py:17]: Umm, do you mean _force_ it on users (by including it in the general import), or _strongly suggest it_ to users by having it at the start of every notebook (and inside the tests)? I think I'd be in favour of the latter actually...

Comment (thomaspinder) [reply gpjax/__init__.py:17]: The latter - it's consistent with how we encourage people to use float64 without enforcing it.

Comment (st--) [reply gpjax/__init__.py]: done

File: gpjax/dataset.py [hunk 27]
```diff
@@ -31,8 +30,8 @@ class Dataset(Pytree):
         y (Optional[Float[Array, "N Q"]]): Output data.
     """
 
-    X: Optional[Float[Array, "N D"]] = None
-    y: Optional[Float[Array, "N Q"]] = None
+    X: Optional[Union[Num[Array, "N D"], Num[Array, "..."]]] = None
+    y: Optional[Union[Num[Array, "N Q"], Num[Array, "..."]]] = None
```
Comment (st--) [inline gpjax/dataset.py:34]: I don't remember why I put it in but the dataset fields should be consistent and always be "N D" / "N Q" even if D or Q are 1?

Comment (st--) [reply gpjax/dataset.py:34]: on second thought it might have been so that it actually gets to the `_check_shape` in the post_init (otherwise beartype errors at the constructor already). maybe let's just leave it that way:
```suggestion
    X: Optional[Num[Array, "N D"]] = None
    y: Optional[Num[Array, "N Q"]] = None
```

Comment (st--) [reply gpjax/dataset.py:34]: does it make sense that they're optional, actually?

File: gpjax/dataset.py [hunk 28]
```diff
@@ -84,7 +83,7 @@ def out_dim(self) -> int:
         return self.y.shape[1]
 
 
-def _check_shape(X: Float[Array, "N D"], y: Float[Array, "N Q"]) -> None:
+def _check_shape(X: Optional[Num[Array, "..."]], y: Optional[Float[Array, "..."]]) -> None:
```
Comment (st--) [inline gpjax/dataset.py:86]:
```suggestion
def _check_shape(X: Optional[Num[Array, "..."]], y: Optional[Num[Array, "..."]]) -> None:
```

**Comment by st--** [comment]
@thomaspinder ok I think it's ready to merge now. I hope I didn't accidentally break anything along the way :sweat_smile:
